Makefile

```diff
@@ -235,24 +240,52 @@ kind-regenerate-controller: manifests generate kustomize $(KIND_KUBECONFIG) ## R
 
-DEPLOYMENT_TARGETS := kind-deploy-registration-operator-hub kind-deploy-registration-operator-managed \
-											kind-approve-cluster kind-load-image kind-regenerate-controller
+OCM_PREP_TARGETS := kind-deploy-registration-operator-hub kind-deploy-registration-operator-managed kind-approve-cluster
+.PHONY: kind-prep-ocm
+kind-prep-ocm: $(OCM_PREP_TARGETS) ## Install OCM registration pieces and connect the clusters
+
 .PHONY: kind-deploy-controller
-kind-deploy-controller: $(DEPLOYMENT_TARGETS) ## Deploy the policy-addon-controller to the kind cluster.
+kind-deploy-controller: kind-prep-ocm kind-load-image kind-regenerate-controller ## Deploy the policy-addon-controller to the kind cluster.
 
 GINKGO = $(LOCAL_BIN)/ginkgo
 .PHONY: e2e-dependencies
 e2e-dependencies: ## Download ginkgo locally if necessary.
 	$(call go-get-tool,github.com/onsi/ginkgo/v2/ginkgo@$(shell awk '/github.com\/onsi\/ginkgo\/v2/ {print $$2}' go.mod))
 
 .PHONY: e2e-test
-e2e-test: e2e-dependencies
-	$(GINKGO) -v --label-filter="!hosted-mode" --fail-fast --slow-spec-threshold=10s test/e2e
+e2e-test: e2e-dependencies ## Run E2E tests.
+	$(GINKGO) -v --label-filter="!hosted-mode" --fail-fast --slow-spec-threshold=10s $(E2E_TEST_ARGS) test/e2e
 
 .PHONY: e2e-test-hosted-mode
 e2e-test-hosted-mode: e2e-dependencies
 	$(GINKGO) -v --label-filter="hosted-mode" --fail-fast --slow-spec-threshold=10s test/e2e
 
+.PHONY: e2e-test-coverage
+e2e-test-coverage: E2E_TEST_ARGS = --json-report=report_e2e.json --output-dir=.
+e2e-test-coverage: e2e-run-instrumented e2e-test e2e-stop-instrumented ## Run E2E tests using instrumented controller.
+
+.PHONY: e2e-build-instrumented
+e2e-build-instrumented:
+	go test -covermode=atomic -coverpkg=$(shell cat go.mod | head -1 | cut -d ' ' -f 2)/... -c -tags e2e ./ -o build/_output/bin/$(IMG)-instrumented
+
+COVERAGE_E2E_OUT ?= coverage_e2e.out
+.PHONY: e2e-run-instrumented
+e2e-run-instrumented: e2e-build-instrumented
+	-KUBECONFIG=$(KIND_KUBECONFIG) kubectl create ns $(CONTROLLER_NAMESPACE)
+	CONFIG_POLICY_CONTROLLER_IMAGE="$(REGISTRY)/config-policy-controller:$(TAG)" \
+	  KUBE_RBAC_PROXY_IMAGE="registry.redhat.io/openshift4/ose-kube-rbac-proxy:v4.10" \
+	  GOVERNANCE_POLICY_SPEC_SYNC_IMAGE="$(REGISTRY)/governance-policy-spec-sync:$(TAG)" \
+	  GOVERNANCE_POLICY_STATUS_SYNC_IMAGE="$(REGISTRY)/governance-policy-status-sync:$(TAG)" \
+	  GOVERNANCE_POLICY_TEMPLATE_SYNC_IMAGE="$(REGISTRY)/governance-policy-template-sync:$(TAG)" \
+	  ./build/_output/bin/$(IMG)-instrumented -test.v -test.run="^TestRunMain$$" -test.coverprofile=$(COVERAGE_E2E_OUT) \
+	  --kubeconfig="$(KIND_KUBECONFIG)" &>build/_output/controller.log &
+
+.PHONY: e2e-stop-instrumented
+e2e-stop-instrumented:
+	ps -ef | grep '$(IMG)' | grep -v grep | awk '{print $$2}' | xargs kill -s SIGUSR1
+	sleep 5 # wait for tests to gracefully shut down
+	-ps -ef | grep '$(IMG)' | grep -v grep | awk '{print $$2}' | xargs kill
```

Comment:
Nit/Optional: Will `sleep 5` always be enough time? Should `main_test.go` create an empty file or set an env instead that the `e2e-stop-instrumented` can read?

Reply:
The second `kill` is just to make sure the target *always* stops the test process. I never actually noticed any appreciable lag time between sending the first signal and the test completing. I just really don't want to leave any unnecessary background processes hanging around.

It might be possible to check that the coverage file was populated, and use that as another signal? But I think that could be error-prone.